src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -18,30 +18,234 @@ namespace mlpack {
+   * its lifetime from other events.
+   *
+   * @param[in] intervals Is a `2 x m` matrix, where each of the m columns has
+   *                      the start sample and the end sample of an interval.
+   *                      This matrix is a space-efficient form of the
+   *                      indicator matrix.
+   * @param[in] indices Indices of the average uniqueness that is returned.
+   *                    This parameter is used for optimization purposes so
+   *                    that in the unit tests the average uniqueness for
+   *                    all events can be accessed while in the production
+   *                    code only a single element is accessed.
+   * @return The average uniqueness of the events in @p intervals.
+   */
+  static arma::vec ComputeAverageUniqueness(
+    const IndMatType& intervals,
+    const arma::uvec& indices,
+    const arma::vec&  concurrency)
+  {
+    arma::vec       avg(indices.n_rows, arma::fill::zeros);
+    const arma::vec invConcurrency(
+      arma::vec(concurrency.n_rows, arma::fill::ones) / concurrency);
+
+    // In production code this loop is only entered once because
+    // then indices.n_rows == 1.
+    for (arma::uword i(0); i < indices.n_rows; ++i) {
+      const arma::uword start(intervals(0, indices[i]));
+      const arma::uword end(intervals(1, indices[i]));
+
+      avg[i] = arma::accu(invConcurrency.rows(start, end)) /
+        (end - start + 1);
+    }
+
+    return avg;
+  }
+
+  /**
+   * Compute probabilities of the next draw for each observation.
+   *
+   * @param[in] phi List of previously drawn observations.
+   * @param[in] phiSize Number of elements in use in @p phi. To
+   *                    avoid reallocations, @p phi has been
+   *                    overallocated already.
+   * @param[in] concurrency Concurrency of the samples selected
+   *                        by @p phi so far.
+   * @param[in] intervals See ComputeAverageUniqueness().
+   * @return The probabilities for each observation to be drawn in the next
+   *         iteration.
+   */
+  static arma::vec ComputeNextDrawProbabilities(
+    arma::uvec&       phi,
+    arma::uword       phiSize,
+    arma::vec&        concurrency,
+    const IndMatType& intervals)
+  {
+    // phi may have less rows than intervals has columns as
+    // phi is grown in multiple steps.
+    assert(phi.n_rows <= intervals.n_cols && phiSize < phi.n_rows);
+
+    // Only in the test code we want to compute the full vector
+    // of average uniqueness. Here, we are always only interested
+    // in the average uniqueness of the latest value that is
+    // added to phi.
+    const arma::uvec indices(1u, arma::fill::value(phiSize));
+    arma::vec        avg(intervals.n_cols);
+
+    for (arma::uword i(0); i < avg.size(); ++i) {
+      // Temporarily assume another concurrency for the current i.
+      concurrency.subvec(intervals(0, i), intervals(1, i)) += 1.0;
```

Comment:
I'm not following why this is needed---can you explain it?

Reply:
@rcurtin we are doing this because we are assuming in each loop iteration another sample that was added to `phi`. But we are not drawing that new element yet. This relates to your question above about the outer concurrency vector that is initially filled with zeros.

Reply:
Yes, I see, but here in this loop we are saying (intuitively) "compute the average uniqueness if we added sample `i` to `phi`"; as the concurrency is just a vector, shouldn't we instead be saying this?

```
concurrency[phiSize] = number of intervals that contain sample i
```